[Plugin] Source plugins

asutoshpalai opened the pull request:
Adds source plugin. This is in continuation of #4608.

segiddins commented on lib/bundler/dsl.rb:
```diff
@@ -125,11 +125,26 @@ def gem(name, *args)
+    def source(source, *args, &blk)
+      options = args.last.is_a?(Hash) ? args.pop.dup : {}
+      options = normalize_hash(options)
+      if options && options.key?("type")
```
options can't be nil here

segiddins commented on lib/bundler/lockfile_parser.rb:
```diff
@@ -22,14 +22,15 @@ class LockfileParser
       Gem::Version.create("1.0") => [DEPENDENCIES, PLATFORMS, GIT, GEM, PATH].freeze,
       Gem::Version.create("1.10") => [BUNDLED].freeze,
-      Gem::Version.create("1.12") => [RUBY].freeze,
+      Gem::Version.create("1.12") => [RUBY, PLUGIN].freeze,
```
this should be 1.13 for PLUGIN

asutoshpalai replied:
Actually there is a spec where `unknown_sections_in_lockfile` should return `nil`

segiddins commented on lib/bundler/plugin.rb:
```diff
@@ -29,7 +31,7 @@ def install(names, options)
     rescue PluginError => e
       paths.values.map {|path| Bundler.rm_rf(path) } if paths
-      Bundler.ui.error "Failed to install plugin #{name}: #{e.message}\n  #{e.backtrace.join("\n  ")}"
+      Bundler.ui.error "Failed to install plugin #{name}: #{e.message}\n  #{e.backtrace[0]}"
```
why this change?

asutoshpalai replied:
You left a comment before merging the previous PR for this.

segiddins commented on lib/bundler/plugin.rb:
```diff
@@ -71,21 +73,49 @@ def add_command(command, cls)
+    def source(name)
+      raise UnknownSourceError, "Source #{name} not found" unless source? name
+
+      load_plugin index.source_plugin name unless @sources.key? name
```
please use parenthesis here

segiddins commented on lib/bundler/plugin/api.rb:
```diff
@@ -23,16 +23,28 @@ module Bundler
+      # @param [Class] (optional) class that handles the source. If not
+      #                 provided, the `self` class will be used.
+      def self.source(source, cls = self)
+        include Bundler::Plugin::API::Source
```
why does this call include?

segiddins commented on lib/bundler/plugin/api/source.rb:
```diff
@@ -0,0 +1,102 @@
+          File.directory?(install_path)
+        end
+
+        def fetch_gemfiles
```
what does this method do?

segiddins commented on lib/bundler/plugin/api/source.rb:
```diff
@@ -0,0 +1,102 @@
+module Bundler
+  module Plugin
+    class API
+      module Source
```
this looks like its a Class?

asutoshpalai replied:
Sorry, forgot to document this class.

The reason this is module is a point where I was looking for suggestions.

My justification:

Our goal here is to provide some default implementations for many of source plugin methods and these implementations should be overridable so that if something doesn't work well for a source plugin they should be able to change it. Also at the same time all the methods of `API` should also be available.

I can see this being done in two ways:

1. (The one that's implemented here) **Runtime inheritance:** We have the implementations in a `module` (since `class` can't be inherited during runtime) and include that module **when a class registers for source plugin**. The advantage is that the only class that plugin writers will have to deal with is `API` and it does all the magic.

2. **Static inheritance:** We can make `API::Source` a subclass of `API` and the source plugins will inherit `API::Source` and the others can inherit `API` only. The advantage is that there is no confusion of runtime inheritance.

segiddins commented on lib/bundler.rb:
```diff
@@ -38,7 +39,6 @@ module Bundler
   autoload :MatchPlatform,          "bundler/match_platform"
   autoload :Mirror,                 "bundler/mirror"
   autoload :Mirrors,                "bundler/mirror"
-  autoload :Plugin,                 "bundler/plugin"
```
why can't this be autoloaded any more?

asutoshpalai replied:
https://github.com/bundler/bundler/pull/4674/files#diff-507c71857bb3b7292c61afb88df1c880L28
This change made starting of `bundler` without `API::Source` fail

segiddins commented on lib/bundler/lockfile_parser.rb:
```diff
@@ -22,14 +22,15 @@ class LockfileParser
     GIT          = "GIT".freeze
     GEM          = "GEM".freeze
     PATH         = "PATH".freeze
+    PLUGIN       = "PLUGIN".freeze
```
maybe `"PLUGIN SOURCE"`?

segiddins commented on lib/bundler/lockfile_parser.rb:
```diff
@@ -116,14 +117,15 @@ def warn_for_outdated_bundler_version
     def parse_source(line)
       case line
-      when GIT, GEM, PATH
+      when GIT, GEM, PATH, PLUGIN
```
`when *source`

segiddins commented on lib/bundler/plugin.rb:
```diff
@@ -96,11 +127,12 @@ def installed?(plugin)
     # @param [Hash] plugins mapped to their installtion path
-    def save_plugins(plugins)
+    # @param [Array<String>] names of auto added source plugins that can be ignored
+    def save_plugins(plugins, optionals = [])
```
`optional_plugins`

segiddins commented on lib/bundler/plugin/api.rb:
```diff
@@ -51,6 +63,10 @@ def method_missing(name, *args, &blk)
       end
+
+      def respond_to_missing?(name, include_private = false)
+        Bundler.respond_to?(name) || super
```
should also use the second param there

segiddins commented on lib/bundler/plugin/api/source.rb:
```diff
@@ -0,0 +1,157 @@
+      # (as long as they behave as expected).
+      # On overriding `initialize` you should call super first.
+      module Source
+        attr_reader :uri, :options
```
please document the attributes

segiddins commented on lib/bundler/plugin/api/source.rb:
```diff
@@ -0,0 +1,157 @@
+        # Note: If the spec method is overridden, this function is not necessary
+        #
+        # @return [Array<String>] paths of the gemfiles that can be installed
+        def fetch_gemfiles
```
why do plugins need to concern themselves with gem files?

asutoshpalai replied:
My bad, they are `gemspec` files.

I guess I am seeing too much of `Gemfile` that I got confused.

segiddins commented on lib/bundler/plugin/api/source.rb:
```diff
@@ -0,0 +1,157 @@
+
+          Bundler::Index.build do |index|
+            files.each do |file|
+              next unless spec = Bundler.load_gemspec(file)
```
so is `file` a gemfile or a gem spec?

asutoshpalai replied:
`gemspec`, will fix that in next commit

segiddins commented on lib/bundler/plugin/dsl.rb:
```diff
@@ -12,9 +12,12 @@ class PluginGemfileError < PluginError; end
       # They will be handled by method_missing
       [:gemspec, :gem, :path, :install_if, :platforms, :env].each {|m| undef_method m }
 
+      attr_reader :auto_plugins
```
please document this, I'm confused about what it is

segiddins commented on lib/bundler/plugin/api.rb:
```diff
@@ -40,7 +40,7 @@ def self.command(command, cls = self)
       #                 provided, the `self` class will be used.
       def self.source(source, cls = self)
-        include Bundler::Plugin::API::Source
+        cls.include Bundler::Plugin::API::Source
```
this might need to be `cls.send :include`

segiddins commented on lib/bundler/plugin/dsl.rb:
```diff
@@ -12,6 +12,14 @@ class PluginGemfileError < PluginError; end
+      #
+      # These plugins are optional and are not installed when there is conflict
+      # with any other plugin.
       attr_reader :auto_plugins
```
`inferred_plugins`?

segiddins commented on lib/bundler/plugin.rb:
```diff
@@ -128,9 +129,9 @@ def installed?(plugin)
-    def save_plugins(plugins, optional_plugins = [])
-      plugins.each do |name, path|
-        path = Pathname.new path
+    def save_plugins(plugins, paths, optional_plugins = [])
```
please doc `paths`

homu commented:
:umbrella: The latest upstream changes (presumably #4700) made this pull request unmergeable. Please resolve the merge conflicts.

segiddins commented on lib/bundler/plugin.rb:
```diff
@@ -197,7 +197,16 @@ def load_plugin(name)
       # done to avoid conflicts
       path = index.plugin_path(name)
 
+      if insert_index = Bundler.rubygems.load_path_insert_index
```
we should just be `#activate`ing the spec here, no?

segiddins replied:
(or using `spec.add_self_to_load_path`?)

asutoshpalai replied:
I tried to do that. But could find a way to get `spec` here.

Even if I store the spec's locations, `add_self_to_load_path` depended on `full_gem_path`, which itself depended on `source`.

The alternative to hard coded `lib` is, we can store the `full_required_paths` with index.

segiddins replied:
we have to do something, since it is totally valid for a gem to have a require path that isn't lib

asutoshpalai replied:
Yes, I am working on extending the serializer to with array. As soon as I get around that, will fix this.

segiddins commented on lib/bundler/plugin/api.rb:
```diff
@@ -59,6 +59,17 @@ def tmp(*names)
+      # @param [String] directory
+      # @param [Proc] block to execute
+      def chdir(dir, &blk)
+        SharedHelpers.chdir(dir, &blk)
```
could just mix in shared helpers to the API?

allenzhao commented:
@asutoshpalai I was trying to edit `lib/bundler/errors.rb` and found that you've used a duplicated status code 23 in L44 which was already taken in L69

asutoshpalai commented:
@allenzhao Yes, right! Sorry, will fix that with the next push!

segiddins commented:
@asutoshpalai @indirect is this ready to merge?

asutoshpalai commented:
@segiddins I think it's now ready to merge

segiddins commented:
@asutoshpalai if you can get this passing, ill give it one last look over and then likely merge

asutoshpalai commented:
This shall pass, there is some problem with travis. It is giving error.

Since the last time it was passing, I just added some new specs that are
passing. (and another small change that I reverted)
On 01-Jul-2016 5:13 pm, "Samuel Giddins" <notifications@github.com> wrote:

> @asutoshpalai <https://github.com/asutoshpalai> if you can get this
> passing, ill give it one last look over and then likely merge
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bundler/bundler/pull/4674#issuecomment-229926398>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AIFEgs_HHllKXcje3g66rk0I2-ed-QmAks5qRP1pgaJpZM4I3kK2>
> .
>

asutoshpalai commented:
@segiddins I fixed the problem with the specs setup and they are passing now.

asutoshpalai commented:
Here is the path with which I fixed the specs
```diff
diff --git a/spec/support/rubygems_ext.rb b/spec/support/rubygems_ext.rb
index 95f596e..e45e69f 100644
--- a/spec/support/rubygems_ext.rb
+++ b/spec/support/rubygems_ext.rb
@@ -6,13 +6,15 @@ module Spec
   module Rubygems
     DEPS = begin
       deps = {
-        "fakeweb artifice rack compact_index" => nil,
+        "fakeweb artifice compact_index" => nil,
         "sinatra" => "1.2.7",
         # Rake version has to be consistent for tests to pass
         "rake" => "10.0.2",
         # 3.0.0 breaks 1.9.2 specs
         "builder" => "2.1.2",
         "bundler" => "1.12.0",
+        # 2.0.1 requires ruby >= 2.2.2
+        "rack" => "1.6.4",
       }
       # ruby-graphviz is used by the viz tests
       deps["ruby-graphviz"] = nil if RUBY_VERSION >= "1.9.3"
@@ -33,7 +35,7 @@ module Spec
         FileUtils.rm_rf(Path.base_system_gems)
         FileUtils.mkdir_p(Path.base_system_gems)
         puts "installing gems for the tests to use..."
-        DEPS.each {|n, v| install_gem(n, v) }
+        DEPS.sort {|a, _| a[1].nil? ? 1 : -1 }.each {|n, v| install_gem(n, v) }
         File.open(manifest_path, "w") {|f| f << manifest.join }
       end
```

homu commented:
:umbrella: The latest upstream changes (presumably #4744) made this pull request unmergeable. Please resolve the merge conflicts.

segiddins commented:
@asutoshpalai I'll merge if you can rebase to get rid of the merge commit :)

segiddins commented:
@homu r+

homu commented:
:pushpin: Commit 57e8172 has been approved by `segiddins`

<!-- @homu r=segiddins 57e817290335570abc1aacdf778e255477403302 -->

homu commented:
:hourglass: Testing commit 57e8172 with merge cc4df62...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/bundler/bundler/builds/142478514)
